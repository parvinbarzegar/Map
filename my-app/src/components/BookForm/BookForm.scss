.book-form {
    display: flex;
    flex-direction: column;
    &__title {
        clip: rect(0 0 0 0);
  clip-path: inset(50%);
  height: 1px;
  overflow: hidden;
  position: absolute;
  white-space: nowrap;
  width: 1px;
    }
    &__input-label{
        text-align: left;
        margin-bottom: 1.5em;
    }
    &__input-control{
        box-sizing: border-box;
    background-color: #f5f8fa;
    display: inline-block;
    width: 90%;
    max-width: 500px;
    height: 40px;
    padding: 9px 10px;
    font-family: "Helvetica Neue",Helvetica,Arial,sans-serif;
    font-size: 16px;
    font-weight: normal;
    line-height: 22px;
    color: #33475b;
    border: 1px solid #cbd6e2;
    }
    &__submit {
        color:blue;
    }
    &__message {
        &--warning{
            color:yellow;
        }
        &--error{
            color:red;
        }
    }
}