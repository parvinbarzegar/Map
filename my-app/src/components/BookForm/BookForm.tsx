import React, { useState, useEffect } from "react";
import DatePicker from "react-datepicker";
import {
  startOfToday,
  startOfTomorrow,
  addDays,
  differenceInDays,
} from "date-fns";
import Button from "../Button/Button";
import Notification from "../Notification/Notification";
import "react-datepicker/dist/react-datepicker.css";
import "./BookForm.scss";
interface IBookFormProps {
  onFinish: Function;
}
const BookForm: React.FunctionComponent<IBookFormProps> = ({ onFinish }) => {
  const [startDate, setStartDate] = useState<Date>(startOfToday());
  const [endDate, setEndDate] = useState<Date>(startOfTomorrow());
  const [boatName, setBoatName] = useState<string>("");
  const [boatNameError, setBoatNameError] = useState<string>("");
  const [boatlengthError, setBoatLengthError] = useState<string>("");
  const [endDateError, setEndDateError] = useState<string>("");
  const [startDateError, setStartDateError] = useState<string>("");
  const [boatLength, setBoatLength] = useState<number>(0);

  useEffect(() => {
    if (differenceInDays(endDate, startDate) < 1) {
      setEndDate(addDays(startDate, 1));
    }
  }, [startDate]);

  const handleClik = () => {
    setBoatNameError("");
    setBoatLengthError("");
    setStartDateError("");
    setEndDateError("");
    if (boatName.length === 0) {
      setBoatNameError("Enter boat name");
    }
    if (boatLength === 0) {
      setBoatLengthError("Enter boat length")
    }
    if (startDate === null) {
      setStartDateError("Select a start date")
    }
    if (endDate === null) {
      setEndDateError("Select an end date")
    } 
    else {
      setBoatNameError("");
      setBoatLengthError("");
      setStartDateError("");
      setEndDateError("");
      onFinish();
    }
  };
  return (
    <div>
      <form className="book-form">
        <legend className="book-form book-form__title">Book your harbor</legend>
        <label className="book-form book-form__input-label">
          Boat name*
          <input
            className="book-form book-form__input-control"
            defaultValue={boatName}
            onChange={(e) => setBoatName(e.target.value)}
            type="text"
            minLength={1}
            maxLength={20}
          />
          {boatNameError && (
            <span role="alert" className="book-form book-form__message--error">
              {boatNameError}
            </span>
          )}
        </label>
        <label className="book-form book-form__input-label">
          Boat length*
          <input
            className="book-form book-form__input-control"
            defaultValue={boatLength}
            onChange={(e) => setBoatLength(Number(e.target.value))}
            type="number"
            min={0}
            max={100}
          />
          {boatlengthError && (
            <span role="alert" className="book-form book-form__message--error">
              {boatlengthError}
            </span>
          )}
        </label>
        <label className="book-form book-form__input-label">
          Start Date*
          <DatePicker
            selected={startDate}
            minDate={startOfToday()}
            onChange={(date: Date) => setStartDate(date)}
            className="book-form book-form__input-control"
          />
          {startDateError && (
            <span role="alert" className="book-form book-form__message--error">
              {startDateError}
            </span>
          )}
        </label>
        <label className="book-form book-form__input-label">
          End Date*
          <DatePicker
            selected={endDate}
            minDate={addDays(startDate, 1)}
            onChange={(date: Date) => setEndDate(date)}
            className="book-form book-form__input-control"
          />
          {endDateError && (
            <span role="alert" className="book-form book-form__message--error">
              {endDateError}
            </span>
          )}
        </label>
        <Button type="primary" onClick={handleClik}>
          Submit
        </Button>
      </form>
    </div>
  );
};
export default BookForm;
