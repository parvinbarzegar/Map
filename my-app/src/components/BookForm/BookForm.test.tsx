import React from 'react';
import { screen, render, cleanup, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import BookForm from './BookForm';

describe('test suits for <BookForm/>',()=> {
    beforeEach(()=> {
        cleanup();
    })
    it('should match to the snapshot',()=> {
        const { container } = render(<BookForm onFinish={()=> jest.fn()}/>);
        expect(container).toMatchSnapshot();
    });
    test('boat name has a correct length', async ()=> {
        render(<BookForm onFinish={()=> jest.fn()}/>);
        const boatName = 'very very very verly loooooooooooong boat nammmmmmeeeeeeeeeeeeee';
        const input = screen.getByLabelText('Boat name');
        expect(input).toHaveProperty("maxLength", 20);
        expect(input).toHaveProperty("minLength",1);
        await userEvent.type(input, boatName);
        screen.getByDisplayValue(boatName.slice(0, 20));
    });
    test('boat name validation',async()=>{
        render(<BookForm onFinish={()=> jest.fn()}/>);
        const boatNameInput = screen.getByLabelText('Boat name');
        const submitButton = screen.getByRole('button');
        await userEvent.click(submitButton);
        expect(screen.getByRole('alert')).toBeInTheDocument();
        userEvent.type(boatNameInput,'boat');
        await userEvent.click(submitButton);
        expect(screen.queryByRole('alert')).not.toBeInTheDocument();
    })
})