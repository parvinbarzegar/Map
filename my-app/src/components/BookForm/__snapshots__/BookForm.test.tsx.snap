// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test suits for <BookForm/> should match to the snapshot 1`] = `
<div>
  <div>
    <form
      class="book-form"
    >
      <legend
        class="book-form book-form__title"
      >
        Book your harbor
      </legend>
      <label
        class="book-form book-form__input-label"
      >
        Boat name
        <input
          class="book-form book-form__input-control"
          maxlength="20"
          minlength="1"
          type="text"
          value=""
        />
        
      </label>
      <label
        class="book-form book-form__input-label"
      >
        Boat length
        <input
          class="book-form book-form__input-control"
          max="100"
          min="0"
          type="number"
          value="0"
        />
      </label>
      <label
        class="book-form book-form__input-label"
      >
        Start Date
        <div
          class="react-datepicker-wrapper"
        >
          <div
            class="react-datepicker__input-container"
          >
            <input
              class=""
              type="text"
              value="01/30/2022"
            />
          </div>
        </div>
      </label>
      <label
        class="book-form book-form__input-label"
      >
        End Date
        <div
          class="react-datepicker-wrapper"
        >
          <div
            class="react-datepicker__input-container"
          >
            <input
              class=""
              type="text"
              value="01/31/2022"
            />
          </div>
        </div>
      </label>
      <button
        class="button button--secondry"
      >
        Submit
      </button>
    </form>
  </div>
</div>
`;
